File: androidApp/src/main/java/com/example/ula_app/android/ui/game/SettingScreen.kt
package com.example.ula_app.android.ui.game

import android.util.Log
import android.widget.Toast
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.text.KeyboardOptions
import androidx.compose.material.Button
import androidx.compose.material.ButtonDefaults
import androidx.compose.material.MaterialTheme
import androidx.compose.material.Tab
import androidx.compose.material.TabRow
import androidx.compose.material.Text
import androidx.compose.material.TextField
import androidx.compose.material3.Slider
import androidx.compose.material3.SliderDefaults
import androidx.compose.material.Switch
import androidx.compose.material.SwitchDefaults
import androidx.compose.runtime.Composable
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.input.KeyboardType
import androidx.compose.ui.text.input.TextFieldValue
import androidx.compose.ui.unit.dp
import com.example.ula_app.android.ULAApplication
import com.example.ula_app.android.ui.component.SettingFieldsRow
import com.example.ula_app.android.ui.viewmodel.UserPreferencesViewModel
import com.example.ula_app.util.DateTimeUtil
import java.io.IOException

private const val TAG = "SettingScreen"

enum class SettingAdvancedSegment(){
    SettingAdvanced,
}

@Composable
fun SettingScreen() {
    val context = LocalContext.current

    val userPreferencesViewModel: UserPreferencesViewModel = ULAApplication.getInstance<UserPreferencesViewModel>()
    val userPreUiState by userPreferencesViewModel.userPreferencesState.collectAsState()

    var currentScreen by remember {
        mutableStateOf("")
    }

    var stepCountSwitchON by remember {
        mutableStateOf(userPreUiState.displaySteps)
    }

    var progressMonsterSwitchON by remember {
        mutableStateOf(userPreUiState.displayMonster)
    }



    var effectiveDays by remember {
        mutableStateOf(userPreUiState.effectiveDays)
    }

    var effectiveDate by remember {
        mutableStateOf(userPreUiState.effectiveDate)
    }

    /*var selectedGoalDaily by remember {
        mutableStateOf(userPreUiState.goal)
    }*/

/*    var goalOptionExpanded by remember {
        mutableStateOf(false)
    }*/

    // TODO: wait to add features--------------------------------------------------------------------------------------------
    /*var selectedGoalWeekly by remember {
        mutableStateOf(0)
    }*/

    var tabIndex by remember { mutableStateOf(0)}
    val tabTitles = listOf("Daily Goal", "Weekly Goal")

    var sliderValueDaily by remember {
        mutableStateOf(userPreUiState.dailyGoal.toFloat())
    }

    var sliderValueWeekly by remember {
        mutableStateOf(userPreUiState.weeklyGoal.toFloat())
    }


    when(currentScreen) {
        SettingAdvancedSegment.SettingAdvanced.name -> {
            SettingAdvancedScreen(
                onBackClicked = {
                    currentScreen = ""
                }
            )
        }

        else -> {
            Column(
                modifier = Modifier
                    .fillMaxSize()
                    .padding(15.dp)
            ) {

                SettingFieldsRow(
                    text = "Display Step Counts",
                    input = {

                        Switch(
                            checked = stepCountSwitchON,
                            onCheckedChange = { switchState ->
                                stepCountSwitchON = switchState
//                    userPreferencesViewModel.setDisplaySteps(true)
                            },
                            colors = SwitchDefaults.colors(
                                checkedThumbColor = MaterialTheme.colors.onPrimary,
                                checkedTrackColor = MaterialTheme.colors.onSecondary,
                                uncheckedThumbColor = Color.Gray,
                                uncheckedTrackColor = Color.DarkGray
                            )
                        )
                    }
                )

                SettingFieldsRow(
                    text = "Display Character",
                    input = {
                        Column(
                            verticalArrangement = Arrangement.Center,
                            horizontalAlignment = Alignment.End
                        ){
                            Switch(
                                checked = progressMonsterSwitchON,
                                onCheckedChange = { switchState ->
                                    progressMonsterSwitchON = switchState
                                },
                                colors = SwitchDefaults.colors(
                                    checkedThumbColor = MaterialTheme.colors.onPrimary,
                                    checkedTrackColor = MaterialTheme.colors.onSecondary,
                                    uncheckedThumbColor = Color.Gray,
                                    uncheckedTrackColor = Color.DarkGray
                                )
                            )

                            Text(
                                text = if (progressMonsterSwitchON) "Monster" else "Human",
                                style = MaterialTheme.typography.h4
                            )

                        }
                    }
                )


                SettingFieldsRow(
                    text = "Effective Days",
                    input = {
                        TextField(
                            modifier = Modifier
                                .width(80.dp),
                            value = TextFieldValue(effectiveDays.toString()),
                            keyboardOptions = KeyboardOptions(
                                keyboardType = KeyboardType.Number
                            ),
                            onValueChange = { newText ->
                                effectiveDays = newText.text.toInt()
                            },
                            enabled = DateTimeUtil.getDayDifference(DateTimeUtil.nowInInstant(), effectiveDate) >= effectiveDays
                        )
                    }
                )

                Column(
                    modifier = Modifier
                        .padding(15.dp)
                        .fillMaxWidth(),
//                    horizontalAlignment = Alignment.CenterHorizontally
                ) {
                    Text(
                        modifier = Modifier.width(200.dp),
                        text = "Change Goal",
                        style = MaterialTheme.typography.h4
                    )

                    Column(
                        modifier = Modifier
                            .padding(20.dp)
                    ) {
                        TabRow(
                            selectedTabIndex = tabIndex,
                            backgroundColor = Color.White,
                            contentColor = Color.Black
                        ) {
                            tabTitles.forEachIndexed {index, title ->
                                Tab(
                                    selected = tabIndex == index,
                                    onClick = {
                                        tabIndex = index
                                    },
                                    text = {
                                        Text(
                                            text = title,
                                            fontWeight = FontWeight.Bold,
                                            style = MaterialTheme.typography.caption
                                        )
                                    }
                                )
                            }
                        }

                        Spacer(modifier = Modifier.height(40.dp))

                        when (tabIndex) {
                            0 -> {
                                Column(
                                    horizontalAlignment = Alignment.CenterHorizontally
                                ) {
                                    Slider(
                                        // TODO: Daily are using two state variables, use only one in the future
                                        value = sliderValueDaily,
                                        onValueChange = { sliderValue_ ->
                                            sliderValueDaily = sliderValue_
                                        },
//                                        onValueChangeFinished = {
//                                            // this is called when the user completed selecting the value
//                                            selectedGoalDaily = sliderValueDaily.toInt()
//                                            Log.d("MainActivity", "sliderValue = $sliderValueDaily")
//                                        },
                                        valueRange = 5000f..30000f,
                                        steps = 4,
                                        colors = SliderDefaults.colors(
                                            thumbColor = MaterialTheme.colors.onPrimary,
                                            activeTrackColor = MaterialTheme.colors.onSecondary
                                        ),
                                        enabled = DateTimeUtil.getDayDifference(DateTimeUtil.nowInInstant(), effectiveDate) >= effectiveDays
                                    )

                                    Text(text = sliderValueDaily.toInt().toString())

                                }
                            }
                            1 -> {
                                Column(
                                    horizontalAlignment = Alignment.CenterHorizontally
                                ) {
                                    Slider(
                                        // TODO: No matter the value of selectedGoalWeekly, it always at the start of the slider, fix it.
                                        value = sliderValueWeekly,
                                        onValueChange = { sliderValue_ ->
                                            sliderValueWeekly = sliderValue_
                                        },
//                                        onValueChangeFinished = {
//                                            // this is called when the user completed selecting the value
//                                            selectedGoalWeekly = sliderValueWeekly.toInt()
//                                            Log.d("MainActivity", "sliderValue = $sliderValueWeekly")
//                                        },
                                        valueRange = 20000f..100000f,
                                        steps = 7,
                                        colors = SliderDefaults.colors(
                                            thumbColor = MaterialTheme.colors.onPrimary,
                                            activeTrackColor = MaterialTheme.colors.onSecondary
                                        ),
                                        enabled = DateTimeUtil.getDayDifference(DateTimeUtil.nowInInstant(), effectiveDate) >= effectiveDays
                                    )

                                    Text(text = sliderValueWeekly.toInt().toString())

                                }
                            }
                        }
                    }

                    /*Dropdown(
                        selectedIndex = changeGoal.toString(),
                        expand = goalOptionExpanded,
                        dropdownTitle = "",
                        dropdownOptions = DataSource.stepOptions,
                        onDropdownClicked = {
                            goalOptionExpanded = it
                        },
                        onDropdownItemClicked = {
                            changeGoal = it.toInt()
                        },
                        enable = DateTimeUtil.getDayDifference(DateTimeUtil.getCurrentDateTime(), effectiveDate) > effectiveDays.text.toLong()
                    )*/

                }

                Row(
                    modifier = Modifier
                        .padding(15.dp)
                        .fillMaxWidth(),
                    horizontalArrangement = Arrangement.SpaceBetween
                ) {
                    // advanced settings button
                    Button(
                        onClick = {
                            currentScreen = SettingAdvancedSegment.SettingAdvanced.name
                        },
                        colors = ButtonDefaults.outlinedButtonColors(contentColor = Color.Black),
                    ){
                        Text(
                            text = "Advanced Settings",
                            style = MaterialTheme.typography.caption
                        )
                    }


                    // Save button
                    Button(
                        onClick = {
                            try {
                                userPreferencesViewModel.setDisplaySteps(stepCountSwitchON)
                                userPreferencesViewModel.setDisplayMonster(progressMonsterSwitchON)
                                userPreferencesViewModel.setEffectiveDays(effectiveDays)
                                userPreferencesViewModel.setEffectiveDate(effectiveDate)
//                                userPreferencesViewModel.setGoal(selectedGoalDaily)
                                userPreferencesViewModel.setDailyGoal(sliderValueDaily.toInt())
                                userPreferencesViewModel.setWeeklyGoal(sliderValueWeekly.toInt())
                                Log.i("SettingsScreen", "Effective?: ${
                                    DateTimeUtil.getDayDifference(
                                        DateTimeUtil.nowInInstant(), effectiveDate) > effectiveDays}")
                                Log.i("SettingsScreen", "EffectiveDate: ${effectiveDate}")
                                Toast.makeText(
                                    context,
                                    "Setting saved successfully",
                                    Toast.LENGTH_LONG,
                                ).show()
                            } catch (e: IOException) {
                                Toast.makeText(
                                    context,
                                    "Setting saved failure",
                                    Toast.LENGTH_LONG,
                                ).show()
                            }


                        },
                        colors = ButtonDefaults.outlinedButtonColors(contentColor = Color.Black),
                    ){
                        Text(
                            text = "Save!",
                            style = MaterialTheme.typography.caption
                        )
                    }
                }



            }
        }
    }
}

File: iosApp/iosApp/game/ui/SettingScreen.swift
//
//  SettingScreen.swift
//  iosApp
//
//  Created by Hao Wu on 9/5/23.
//  Copyright © 2023 orgName. All rights reserved.
//

import Foundation
import SwiftUI
import shared

struct SettingScreen: View {
    
    @ObservedObject var settingState: SettingState
    
    let dailyGoalOption: [Int] = [
        5000,
        10000,
        15000,
        20000
    ]
    
    let weeklyGoalOption: [Int] = [
        20000,
        30000,
        40000,
        50000,
        60000,
        70000,
        80000,
        90000,
        100000
    ]
    
    init() {
        self.settingState = SettingState()
        
    }
    
    // TODO: UI needs fix
    var body: some View {
        
        NavigationView {
            
            VStack(alignment: .leading, spacing: 20.0) {
                
                // Display Step Counts
                HStack {
                    Toggle("Display Step Counts", isOn: $settingState.stepCountSwitchON)
                        .toggleStyle(SwitchToggleStyle(tint: .pink))
                }
                
                // Display Character
                HStack {
                    Toggle(isOn: $settingState.progressMonsterSwitchON) {
                        
                        if settingState.progressMonsterSwitchON {
                            Text("Display Character(Monster)")
                        } else {
                            Text("Display Character(Human)")
                        }
                       
                    }
                    .toggleStyle(SwitchToggleStyle(tint: .pink))
                }
                
                // Effective Days
                HStack{
                    Text("Effective Days")
                    Spacer()
                    TextField("Enter your name", value: $settingState.effectiveDays, formatter: NumberFormatter())
                        .textFieldStyle(RoundedBorderTextFieldStyle())
                        .frame(width: 50)
//                        .disabled(DateTimeUtil.getDayDifference(next: DateTimeUtil.getCurrentDateTime(), previous: effectiveDate) >= effectiveDays)
                }
                
                // Change Goal
                VStack{
                    
                    HStack{
                        Text("Change Goal")
                        Spacer()
                        
                        
                    }.frame(height: 20)
                    
                    VStack {
                        VStack(alignment: .center) {
                            Tabs(
                                tabs: .constant(["Daily Goal", "Weekly Goal"]),
                                selection: $settingState.tabIndex,
                                underlineColor: .black) { title, isSelected in
                                    Text(title.uppercased())
                //                       .font(.system(size: 20))
                                       .fontWeight(.semibold)
                                       .foregroundColor(isSelected ? .black : .gray)
                                }

                            Text(settingState.tabIndex == 0 ? "\(Int(settingState.dailyGoal))" : "\(Int(settingState.weeklyGoal))").padding(.top, 20)
                            Slider(
                                value: settingState.tabIndex == 0 ? $settingState.dailyGoal : $settingState.weeklyGoal,
                                in: settingState.tabIndex == 0 ? 5000.0...30000 : 20000.0...100000.0,
                                step: settingState.tabIndex == 0 ? 5000.0 : 10000.0
                            )
                            .tint(Color.pink)
                            .padding(.horizontal, 20)
                            Spacer()
                        }
                        .padding(.top, 20)
                    }
                    
                    Spacer()
                }
                
                
                
                
                // Navigation button to the advanced settings
                NavigationLink(destination: SettingAdvancedScreen()) {
//                    Button("Advanced Settings") {}
                    Text("Advanced Settings")
                        .foregroundStyle(.white)
                        .padding(12)
                        .background(
                            RoundedRectangle(cornerRadius: 10)
                                .fill(Color.gray)
//                                .shadow(color: .gray, radius: 2, x: 0, y: 2)
                        )
                        .clipShape(Rectangle())
                }
                
                // Button to save all setting changes
                Button("Save!") {
                    settingState.writeDisplaySteps(displaySteps: settingState.stepCountSwitchON)
                    settingState.writeDisplayMonster(displayMonster: settingState.progressMonsterSwitchON)
                    settingState.writeEffectiveDate(effectiveDate: settingState.effectiveDate)
                    settingState.writeEffectiveDays(effectiveDays: settingState.effectiveDays)
                    settingState.writeDailyGoal(dailyGoal: Int(settingState.dailyGoal))
                    settingState.writeWeeklyGoal(weeklyGoal: Int(settingState.weeklyGoal))
                }
                .padding(12)
                .foregroundStyle(.white)
                .background(
                    RoundedRectangle(cornerRadius: 10)
                        .fill(Color.gray)
//                        .shadow(color: .gray, radius: 2, x: 0, y: 2)
                )
                .clipShape(Rectangle())
                
            }
                .padding()
                .navigationBarTitle("Settings")
                .frame(alignment: .topLeading)
            
        }
        
    }
    
}


struct SettingScreen_Previews: PreviewProvider {
    static var previews: some View {
        SettingScreen()
    }
}
